File: app/src/main/java/org/monksanctum/xand11/comm/Request.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.comm;

import org.monksanctum.xand11.extension.Extension;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

public class Request {
    public static final byte CREATE_WINDOW = 1;
    public static final byte CHANGE_WINDOW_ATTRIBUTES = 2;
    public static final byte GET_WINDOW_ATTRIBUTES = 3;
    public static final byte DESTROY_WINDOW = 4;
    public static final byte DESTROY_SUBWINDOWS = 5;
    public static final byte CHANGE_SAVE_SET = 6;
    public static final byte REPARENT_WINDOW = 7;
    public static final byte MAP_WINDOW = 8;
    public static final byte MAP_SUBWINDOWS = 9;
    public static final byte UNMAP_WINDOW = 10;
    public static final byte UNMAP_SUBWINDOWS = 11;
    public static final byte CONFIGURE_WINDOW = 12;
    public static final byte CIRCULATE_WINDOW = 13;
    public static final byte GET_GEOMETRY = 14;
    public static final byte QUERY_TREE = 15;

    public static final byte INTERN_ATOM = 16;
    public static final byte GET_ATOM_NAME = 17;

    public static final byte CHANGE_PROPERTY = 18;
    public static final byte DELETE_PROPERTY = 19;
    public static final byte GET_PROPERTY = 20;
    public static final byte LIST_PROPERTIES = 21;

    public static final byte SET_SELECTION_OWNER = 22;
    public static final byte GET_SELECTION_OWNER = 23;
    public static final byte CONVERT_SELECTION = 24;

    public static final byte SEND_EVENT = 25;

    public static final byte GRAB_POINTER = 26;
    public static final byte UNGRAB_POINTER = 27;
    public static final byte GRAB_BUTTON = 28;
    public static final byte UNGRAB_BUTTON = 29;
    public static final byte CHANGE_ACTIVE_POINTER_GRAB = 30;
    public static final byte GRAB_KEYBOARD = 31;
    public static final byte UNGRAB_KEYBOARD = 32;
    public static final byte GRAB_KEY = 33;
    public static final byte UNGRAB_KEY = 34;
    public static final byte ALLOW_EVENTS = 35;
    public static final byte GRAB_SERVER = 36;
    public static final byte UNGRAB_SERVER = 37;

    public static final byte QUERY_POINTER = 38;
    public static final byte GET_MOTION_EVENTS = 39;
    public static final byte TRANSLATE_COORDINATES = 40;
    public static final byte WRAP_POINTER = 41;
    public static final byte SET_INPUT_FOCUS = 42;
    public static final byte GET_INPUT_FOCUS = 43;

    public static final byte QUERY_KEYMAP = 44;
    public static final byte OPEN_FONT = 45;
    public static final byte CLOSE_FONT = 46;
    public static final byte QUERY_FONT = 47;
    public static final byte QUERY_TEXT_EXTENTS = 48;
    public static final byte LIST_FONTS = 49;
    public static final byte LIST_FONTS_WITH_INFO = 50;
    public static final byte SET_FONT_PATH = 51;
    public static final byte GET_FONT_PATH = 52;

    public static final byte CREATE_PIXMAP = 53;
    public static final byte FREE_PIXMAP = 54;
    public static final byte CREATE_GC = 55;
    public static final byte CHANGE_GC = 56;
    public static final byte COPY_GC = 57;
    public static final byte SET_DASHES = 58;
    public static final byte SET_CLIP_RECTANGLES = 59;
    public static final byte FREE_GC = 60;

    public static final byte CLEAR_AREA = 61;
    public static final byte COPY_AREA = 62;
    public static final byte COPY_PLANE = 63;
    public static final byte POLY_POINT = 64;
    public static final byte POLY_LINE = 65;
    public static final byte POLY_SEGMENT = 66;
    public static final byte POLY_RECTANGLE = 67;
    public static final byte POLY_ARC = 68;
    public static final byte FILL_POLY = 69;
    public static final byte POLY_FILL_RECTANGLE = 70;
    public static final byte POLY_FILL_ARC = 71;

    public static final byte PUT_IMAGE = 72;
    public static final byte GET_IMAGE = 73;
    public static final byte POLY_TEXT_8 = 74;
    public static final byte POLY_TEXT_16 = 75;
    public static final byte IMAGE_TEXT_8 = 76;
    public static final byte IMAGE_TEXT_16 = 77;

    public static final byte CREATE_COLORMAP = 78;
    public static final byte FREE_COLORMAP = 79;
    public static final byte COPY_COLORMAP_AND_FREE = 80;
    public static final byte INSTALL_COLORMAP = 81;
    public static final byte UNINSTALL_COLORMAP = 82;
    public static final byte LIST_INSTALLED_COLORMAP = 83;
    public static final byte ALLOC_COLOR = 84;
    public static final byte ALLOC_NAMED_COLOR = 85;
    public static final byte ALLOC_COLOR_CELLS = 86;
    public static final byte ALLOC_COLOR_PLANES = 87;
    public static final byte FREE_COLORS = 88;
    public static final byte STORE_COLORS = 89;
    public static final byte STORE_NAMED_COLOR = 90;
    public static final byte QUERY_COLORS = 91;
    public static final byte LOOKUP_COLOR = 92;

    public static final byte CREATE_CURSOR = 93;
    public static final byte CREATE_GLYPH_CURSOR = 94;
    public static final byte FREE_CURSOR = 95;
    public static final byte RECOLOR_CURSOR = 96;
    public static final byte QUERY_BEST_SIZE = 97;

    public static final byte QUERY_EXTENSION = 98;
    public static final byte LIST_EXTENSIONS = 99;

    public static final byte CHANGE_KEYBOARD_MAPPING = 100;
    public static final byte GET_KEYBOARD_MAPPING = 101;
    public static final byte CHANGE_KEYBOARD_CONTROL = 102;
    public static final byte GET_KEYBOARD_CONTROL = 103;
    public static final byte BELL = 104;

    public static final byte CHANGE_POINTER_CONTROL = 105;
    public static final byte GET_POINTER_CONTROL = 106;

    public static final byte SET_SCREEN_SAVER = 107;
    public static final byte GET_SCREEN_SAVER = 108;

    public static final byte CHANGE_HOSTS = 109;
    public static final byte LIST_HOSTS = 110;
    public static final byte SET_ACCESS_CONTROL = 111;
    public static final byte SET_CLOSE_DOWN_MODE = 112;
    public static final byte KILL_CLIENT = 113;

    public static final byte ROTATE_PROPERTIES = 114;

    public static final byte FORCE_SCREEN_SAVER = 115;

    public static final byte SET_POINTER_MAPPING = 116;
    public static final byte GET_POINTER_MAPPING = 117;
    public static final byte SET_MODIFIER_MAPPING = 118;
    public static final byte GET_MODIFIER_MAPPING = 119;

    public static final byte NO_OP = 127;

    private static final String[] sNames = new String[256];

    public static void populateNames() {
        Class<Request> cls = Request.class;
        for (Field field : cls.getDeclaredFields()) {
            if (Modifier.isStatic(field.getModifiers()) && field.getType() == byte.class) {
                try {
                    byte index = (byte) field.get(null);
                    sNames[index] = field.getName();
                } catch (IllegalAccessException e) {
                }
            }
        }
    }

    public static String getName(int what) {
        if (sNames[what] == null) {
            return "Unknown(" + what + ")";
        }
        return sNames[what];
    }
}

File: app/src/main/java/org/monksanctum/xand11/comm/Request.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.comm

import org.monksanctum.xand11.extension.Extension

import java.lang.reflect.Field
import java.lang.reflect.Modifier

object Request {
    val CREATE_WINDOW: Byte = 1
    val CHANGE_WINDOW_ATTRIBUTES: Byte = 2
    val GET_WINDOW_ATTRIBUTES: Byte = 3
    val DESTROY_WINDOW: Byte = 4
    val DESTROY_SUBWINDOWS: Byte = 5
    val CHANGE_SAVE_SET: Byte = 6
    val REPARENT_WINDOW: Byte = 7
    val MAP_WINDOW: Byte = 8
    val MAP_SUBWINDOWS: Byte = 9
    val UNMAP_WINDOW: Byte = 10
    val UNMAP_SUBWINDOWS: Byte = 11
    val CONFIGURE_WINDOW: Byte = 12
    val CIRCULATE_WINDOW: Byte = 13
    val GET_GEOMETRY: Byte = 14
    val QUERY_TREE: Byte = 15

    val INTERN_ATOM: Byte = 16
    val GET_ATOM_NAME: Byte = 17

    val CHANGE_PROPERTY: Byte = 18
    val DELETE_PROPERTY: Byte = 19
    val GET_PROPERTY: Byte = 20
    val LIST_PROPERTIES: Byte = 21

    val SET_SELECTION_OWNER: Byte = 22
    val GET_SELECTION_OWNER: Byte = 23
    val CONVERT_SELECTION: Byte = 24

    val SEND_EVENT: Byte = 25

    val GRAB_POINTER: Byte = 26
    val UNGRAB_POINTER: Byte = 27
    val GRAB_BUTTON: Byte = 28
    val UNGRAB_BUTTON: Byte = 29
    val CHANGE_ACTIVE_POINTER_GRAB: Byte = 30
    val GRAB_KEYBOARD: Byte = 31
    val UNGRAB_KEYBOARD: Byte = 32
    val GRAB_KEY: Byte = 33
    val UNGRAB_KEY: Byte = 34
    val ALLOW_EVENTS: Byte = 35
    val GRAB_SERVER: Byte = 36
    val UNGRAB_SERVER: Byte = 37

    val QUERY_POINTER: Byte = 38
    val GET_MOTION_EVENTS: Byte = 39
    val TRANSLATE_COORDINATES: Byte = 40
    val WRAP_POINTER: Byte = 41
    val SET_INPUT_FOCUS: Byte = 42
    val GET_INPUT_FOCUS: Byte = 43

    val QUERY_KEYMAP: Byte = 44
    val OPEN_FONT: Byte = 45
    val CLOSE_FONT: Byte = 46
    val QUERY_FONT: Byte = 47
    val QUERY_TEXT_EXTENTS: Byte = 48
    val LIST_FONTS: Byte = 49
    val LIST_FONTS_WITH_INFO: Byte = 50
    val SET_FONT_PATH: Byte = 51
    val GET_FONT_PATH: Byte = 52

    val CREATE_PIXMAP: Byte = 53
    val FREE_PIXMAP: Byte = 54
    val CREATE_GC: Byte = 55
    val CHANGE_GC: Byte = 56
    val COPY_GC: Byte = 57
    val SET_DASHES: Byte = 58
    val SET_CLIP_RECTANGLES: Byte = 59
    val FREE_GC: Byte = 60

    val CLEAR_AREA: Byte = 61
    val COPY_AREA: Byte = 62
    val COPY_PLANE: Byte = 63
    val POLY_POINT: Byte = 64
    val POLY_LINE: Byte = 65
    val POLY_SEGMENT: Byte = 66
    val POLY_RECTANGLE: Byte = 67
    val POLY_ARC: Byte = 68
    val FILL_POLY: Byte = 69
    val POLY_FILL_RECTANGLE: Byte = 70
    val POLY_FILL_ARC: Byte = 71

    val PUT_IMAGE: Byte = 72
    val GET_IMAGE: Byte = 73
    val POLY_TEXT_8: Byte = 74
    val POLY_TEXT_16: Byte = 75
    val IMAGE_TEXT_8: Byte = 76
    val IMAGE_TEXT_16: Byte = 77

    val CREATE_COLORMAP: Byte = 78
    val FREE_COLORMAP: Byte = 79
    val COPY_COLORMAP_AND_FREE: Byte = 80
    val INSTALL_COLORMAP: Byte = 81
    val UNINSTALL_COLORMAP: Byte = 82
    val LIST_INSTALLED_COLORMAP: Byte = 83
    val ALLOC_COLOR: Byte = 84
    val ALLOC_NAMED_COLOR: Byte = 85
    val ALLOC_COLOR_CELLS: Byte = 86
    val ALLOC_COLOR_PLANES: Byte = 87
    val FREE_COLORS: Byte = 88
    val STORE_COLORS: Byte = 89
    val STORE_NAMED_COLOR: Byte = 90
    val QUERY_COLORS: Byte = 91
    val LOOKUP_COLOR: Byte = 92

    val CREATE_CURSOR: Byte = 93
    val CREATE_GLYPH_CURSOR: Byte = 94
    val FREE_CURSOR: Byte = 95
    val RECOLOR_CURSOR: Byte = 96
    val QUERY_BEST_SIZE: Byte = 97

    val QUERY_EXTENSION: Byte = 98
    val LIST_EXTENSIONS: Byte = 99

    val CHANGE_KEYBOARD_MAPPING: Byte = 100
    val GET_KEYBOARD_MAPPING: Byte = 101
    val CHANGE_KEYBOARD_CONTROL: Byte = 102
    val GET_KEYBOARD_CONTROL: Byte = 103
    val BELL: Byte = 104

    val CHANGE_POINTER_CONTROL: Byte = 105
    val GET_POINTER_CONTROL: Byte = 106

    val SET_SCREEN_SAVER: Byte = 107
    val GET_SCREEN_SAVER: Byte = 108

    val CHANGE_HOSTS: Byte = 109
    val LIST_HOSTS: Byte = 110
    val SET_ACCESS_CONTROL: Byte = 111
    val SET_CLOSE_DOWN_MODE: Byte = 112
    val KILL_CLIENT: Byte = 113

    val ROTATE_PROPERTIES: Byte = 114

    val FORCE_SCREEN_SAVER: Byte = 115

    val SET_POINTER_MAPPING: Byte = 116
    val GET_POINTER_MAPPING: Byte = 117
    val SET_MODIFIER_MAPPING: Byte = 118
    val GET_MODIFIER_MAPPING: Byte = 119

    val NO_OP: Byte = 127

    private val sNames = arrayOfNulls<String>(256)

    fun populateNames() {
        val cls = Request::class.java
        for (field in cls!!.getDeclaredFields()) {
            if (Modifier.isStatic(field.getModifiers()) && field.getType() == Byte::class.javaPrimitiveType) {
                try {
                    val index = field.get(null) as Byte
                    sNames[index.toInt()] = field.getName()
                } catch (e: IllegalAccessException) {
                }

            }
        }
    }

    fun getName(what: Int): String? {
        return if (sNames[what] == null) {
            "Unknown($what)"
        } else sNames[what]
    }
}
